pf/internal/defaults/defaults.go

```diff
@@ -0,0 +1,293 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package defaults
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strconv"
+
+	"github.com/hashicorp/terraform-plugin-log/tflog"
+
+	"github.com/pulumi/pulumi/sdk/v3/go/common/resource"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/contract"
+
+	"github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tfbridge"
+	shim "github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tfshim"
+	"github.com/pulumi/pulumi-terraform-bridge/v3/unstable/propertyvalue"
+)
+
+type ApplyDefaultInfoValuesArgs struct {
+	// Required. The configuration property map to extend with defaults.
+	PropertyMap resource.PropertyMap
+
+	// Toplevel schema map for the resource, data source or provider.
+	TopSchemaMap shim.SchemaMap
+
+	// Toplevel SchemaInfo configuration matching TopSchemaMap.
+	TopFieldInfos map[string]*tfbridge.SchemaInfo
+
+	// Optional. Note that ResourceInstance need not be specified for Invoke or Configure processing.
+	ResourceInstance *tfbridge.PulumiResource
+
+	// Optional. If known, these are the provider-level configuration values, to support DefaultInfo.Config.
+	ProviderConfig resource.PropertyMap
+}
+
+// Transforms a PropertyMap to apply default values specified in DefaultInfo.
+//
+// These values are specified at the bridged provider configuration level, and are applied before any Terraform
+// processing; therefore the function works at Pulumi level (transforming a PropertyMap).
+func ApplyDefaultInfoValues(ctx context.Context, args ApplyDefaultInfoValuesArgs) resource.PropertyMap {
+
+	// Can short-circuit the entire processing if there are no matching SchemaInfo entries, and therefore no
+	// matching DefaultInfo entries at all.
+	if args.TopFieldInfos == nil {
+		return args.PropertyMap
+	}
+
+	t := &defaultsTransform{
+		resourceInstance: args.ResourceInstance,
+		topSchemaMap:     args.TopSchemaMap,
+		topFieldInfos:    args.TopFieldInfos,
+		providerConfig:   args.ProviderConfig,
+	}
+	result := t.withDefaults(ctx, make(resource.PropertyPath, 0), resource.NewObjectProperty(args.PropertyMap))
+	if !result.IsObject() {
+		contract.Failf("defaultsTransform.withDefaults returned a non-object value")
+	}
+	return result.ObjectValue()
+}
+
+func getDefaultValue(
+	ctx context.Context,
+	res *tfbridge.PulumiResource,
+	fieldSchema shim.Schema,
+	defaultInfo *tfbridge.DefaultInfo,
+	providerConfig resource.PropertyMap,
+) (resource.PropertyValue, bool, error) {
+	na := resource.NewNullProperty()
+
+	if defaultInfo == nil {
+		return na, false, nil
+	}
+	if defaultInfo.From != nil && res != nil {
+		raw, err := defaultInfo.From(res)
+		if err != nil {
+			return na, false, err
+		}
+		if raw == nil {
+			return na, false, nil
+		}
+		return recoverDefaultValue(raw), true, nil
+
+	} else if defaultInfo.EnvVars != nil {
```

Comment:
Both for clarity and to match https://github.com/pulumi/pulumi-terraform-bridge/blob/16f3cd4597abb67da5b9a213e5e216e603fb3b26/pkg/tfbridge/schema.go#L617.
```suggestion
	} else if len(defaultInfo.EnvVars) != 0 {
```